DM-21687: Add draft of how-to PipelineTask (#102)
lsst/pipe_base: doc/lsst.pipe.base/creating-a-pipelinetask.rst

            # set dimension cutouts to 3 times the apRad times 2 (for diameter)
            dimensions = (3*self.apRad*2, 3*self.apRad*2)

            # Get indexes for each pixel
            indy, indx = np.indices(dimensions)

            # Loop over each record in the catalog
            for source in inputCatalog:
                # Create an aperture and measure the flux
                center = Point2I(source.getCentroid())
                center = (center.getY(), center.getX())
                # Create a cutout
                stamp = exposure.image.array[center[0]-3*self.apRad: center[0]+3*self.apRad,
                                            center[1]-3*self.apRad: center[1]+3*self.apRad]
                mask = ((indy - center[0])**2
                        + (indx - center[0])**2)**0.5 < self.apRad
                flux = np.sum(stamp*mask)

                # Set the flux field of this source
                source.set(self.apKey, flux)

            return pipeBase.Struct(outputCatalog=self.outputCatalog)

These changes allow you to load in and use schemas to initialize your task before
any actual data is loaded to be passed to the algorithm code located in the

> `run` method. Inside the `run` method, the output catalog copies all the

Reviewer: ```suggestion
``run`` method. Inside the ``run`` method, the output catalog copies all the
```

records from the input into itself. The loop can the go over the output
catalog, and insert the new measurements right into the output catalog.

> One thing to note about `InitInput` connections is that they do not take any

Reviewer: ```suggestion
One thing to note about `~lsst.pipe.base.connectionTypes.InitInput` connections is that they do not take any
```

dimensions. This is because the sort of data loaded will correspond to a
given data-set type produced by a task, and not by (possibly multiple)

> executions of a run method over data-sets that have dimensions, that is to

Reviewer: Find a way to break up the sentence (possibly at the comma).

say these data-sets are unique to the task itself and not tied to the unit of
work that the task operates on.

> In the same way the schema schema from some previous stage of was added, the

Reviewer: ```suggestion
In the same way the input schema from some previous stage of processing was added, the
```
?

output schema for your task should also be persisted so some other

> `PipelineTask` can make use of it. To do this you should add an `InitOutput`

Reviewer: ```suggestion
|PipelineTask| can make use of it. To do this you should add an `~lsst.pipe.base.connectionTypes.InitOutput`
```

                                              "abstract_filter", "skymap")):
        ...
        outputSchema = connectionTypes.InitOutput(doc="Schema created in Aperture PipelineTask",
                                                  storageClass="SourceCatalog",
                                                  name="customAperture_schema")
                                                        
        ...

    class ApertureTask(pipeBase.PipelineTask):

        ...
        
        def __init__(self, config: pexConfig.Config, initInput:Mapping,
                     *args, **kwargs):
            ...
            # Get the output schema
            self.schema = mapper.getOutputSchema()
        
            # Create the output catalog
            self.outputCatalog = afwTable.SourceCatalog(self.schema)

            # Put the outputSchema into a SourceCatalog container. This var name
            # matches an initOut so will be persisted
            self.outputSchema = afwTable.SourceCatalog(self.schema)

In the init method we associate the variable we would like to output with a
name that matches the variable name used in the connection class. The

> activator uses this shared name to know which variable should be persisted.

Reviewer: ```suggestion
activator uses this shared name to know that variable should be persisted.
```

The complete updated example can be found in :ref:`pipeline-appendix-b`.

Optional Data-sets
==================
Sometimes it is useful to have a task that optionally uses a data-set. In the

> case of the example you have been building, task this might be a background

Reviewer: ```suggestion
case of the example task you have been building, this might be a background
```

data-sets.

.. code-block:: python

    class ApertureTaskConnections(pipeBase.PipelineTaskConnections,
                                  dimensions=("visit", "detector", "band")):
        ...
        background = connectionTypes.Input(doc="Background model for the exposure",
                                           storageClass="Background",
                                           name="calexpBackground",
                                           dimensions=("visit", "detector", "band"))
        ...

Now your `PipelineTask` will load the background each time the task is run.
How do you make this optional? First, add a configuration field in your
config class to allow the user to specify if it is to be loaded like thus

.. code-block:: python

    class ApertureTaskConfig(pipeBase.PipelineTaskConfig,
                             pipelineConnections=ApertureTaskConnections):
        ...
        doLocalBackground = pexConfig.Field(doc="Should the background be added "
                                                "before doing photometry",
                                            dtype=bool, default=False)

> The `\_\_init\_\_` method of the connection class is given an instance of the

Reviewer: ```suggestion
The ``__init__`` method of the connection class is given an instance of the
```

> tasks config class after all overrides have been applied. This provides you an

Reviewer: ```suggestion
task's configuration class after all overrides have been applied. This provides you an
```

opportunity change the behavior of the connection class according to various

> config options.

Reviewer: ```suggestion
configuration options.
```

.. code-block:: python

    class ApertureTaskConnections(pipeBase.PipelineTaskConnections,
                                  dimensions=("visit", "detector",
                                              "abstract_filter", "skymap")):
        ...
        background = ct.Input(doc="Background model for the exposure",
                              storageClass="Background",
                              name="calexpBackground",
                              dimensions=("visit", "detector", "abstract_filter",
                                          "skymap"))
        ...

        def __init__(self, *, config=None):
            super().__init__(config=config)
            
            if config.doLocalBackground is False:
                self.inputs.remove("background")

> Your connection class now looks at the value of `doLocalBackground` on the

Reviewer: ```suggestion
Your connection class now looks at the value of ``doLocalBackground`` on the
```

> config object and if it is false, removes it from the connection instances

Reviewer: ```suggestion
``config`` object and if it is ``False``, removes it from the connection instances
```

list of input connections. Connection classes keep track of what connections
are defined in sets. Each set contains the variable names of a connection,
and the sets themselves are identified by the type of connection they
contain. In the example you are modifying the set of input connections. The
names for each of the sets are as follows:

> * initInputs
> * initOutputs
> * inputs
> * prerequisiteInputs
> * outputs 

Reviewer: ```suggestion
* ``initInputs``
* ``initOutputs``
* ``inputs``
* ``prerequisiteInputs``
* ``outputs``
```

> The last step in modifying your task will be to update the run method to take

Reviewer: ```suggestion
The last step in modifying your task will be to update the ``run`` method to take
```

into account that a background may or may not be supplied.

.. code-block:: python

    ...
    import typing

    import lsst.afw.math as afwMath

    ...

    class ApertureTask(pipeBase.PipelineTask):
        
        ...

        def run(self, exposure: afwImage.Exposure,
                inputCatalog: afwTable.SourceCatalog,

>                 background: typing.Optional[afwMath.BackgroundList] = None

Reviewer: Normally we import `Optional` directly from `typing` because it's so commonly used

```
from typing import Optional
```

                ) -> pipeBase.Struct:
            # If a background is supplied, add it back to the image so local
            # background subtraction can be done.
            if background is not None:
                exposure.image.array += background.image

            ...

            # Loop over each record in the catalog
            for source in outputCatalog:

                ...
                distance = ((indy - center[0])**2
                            + (indx - center[0])**2)**0.5
                mask = distance < self.apRad
                flux = np.sum(stamp*mask)

                # Do local background subtraction
                if background is not None:
                    outerAn = distance < 2.5*self.apRad
                    innerAn = distance < 1.5*self.apRad
                    annulus = outerAn - innerAn
                    localBackground = np.mean(exposure.image.array*annulus)
                    flux -= np.sum(mask)*localBackground

                ...
            
            return pipeBase.Struct(outputCatalog=self.outputCatalogs)

> The run method now takes an argument named background, which defaults to a

Reviewer: ```suggestion
The ``run`` method now takes an argument named ``background``, which defaults to a
```
